provider/sidecar.go

```diff
@@ -70,150 +89,169 @@ func (provider *Sidecar) Provide(configurationChan chan<- types.ConfigMessage, p
+		}
+
+		for _, serv := range services {
+			if serv.IsAlive() {
+				ipAddr, err := net.LookupIP(serv.Hostname)
+
+				for i := 0; i < len(serv.Ports); i++ {
+					// TODO: is there any point to add unreachable hosts?
+					var hostname string
+					if err != nil {
+						log.Warn("Failed to resolve IP address: ", err)
+						hostname = serv.Hostname
+					} else {
+						hostname = ipAddr[0].String()
+					}
+
+					backend.Servers[serv.Hostname] = types.Server{
+						URL: fmt.Sprintf("http://%s:%d", hostname, serv.Ports[i].Port),
+					}
+				}
+			}
+		}
+	}
+
 	provider.configurationChan <- types.ConfigMessage{
 		ProviderName:  "sidecar",
-		Configuration: conf,
+		Configuration: config,
 	}
+
+	log.Info("Finished loading sidecar config")
+
 	return nil
 }
 
-func (provider *Sidecar) sidecarWatcher() error {
-	//set timeout to be just a bot more than connection refresh interval
+func (provider *Sidecar) sidecarWatcher() {
+	// Set timeout to be just a bit more than connection refresh interval
 	provider.connTimer = time.NewTimer(time.Duration(provider.RefreshConn))
-	tr := &http.Transport{ResponseHeaderTimeout: 0}
-	client := &http.Client{
-		Timeout:   0,
-		Transport: tr}
+
 	log.Debugf("Using %s Sidecar connection refresh interval", provider.RefreshConn)
-	provider.recycleConn(client, tr)
-	return nil
-}
+	for {
+		// Wrap code in an anonymous function because defer has function scope
+		func() {
+			// Use refresh interval to occasionally reconnect to Sidecar in case the stream connection is lost
+			req, err := http.NewRequest(http.MethodGet, provider.Endpoint+"/watch", nil)
+			if err != nil {
+				log.Errorf("Error creating http request to Sidecar instance '%s': %s", provider.Endpoint, err)
+				time.Sleep(5 * time.Second)
+				return
+			}
 
-func (provider *Sidecar) recycleConn(client *http.Client, tr *http.Transport) {
-	var err error
-	var resp *http.Response
-	var req *http.Request
-	for { //use refresh interval to occasionally reconnect to Sidecar in case the stream connection is lost
-		req, err = http.NewRequest("GET", provider.Endpoint+"/watch", nil)
-		if err != nil {
-			log.Errorf("Error creating http request to Sidecar: %s, Error: %s", provider.Endpoint, err)
-			continue
-		}
-		resp, err = client.Do(req)
-		if err != nil {
-			log.Errorf("Error connecting to Sidecar: %s, Error: %s", provider.Endpoint, err)
-			time.Sleep(5 * time.Second)
-			continue
-		}
-		safe.Go(func() { catalog.DecodeStream(resp.Body, provider.callbackLoader) })
+			cx, cancel := context.WithCancel(context.Background())
+			// Cancel the infinite timeout request automatically after we reset connTimer
+			defer cancel()
 
-		//wait on refresh connection timer.  If this expires we haven't seen an update in a
-		//while and should cancel the request, reset the time, and reconnect just in case
-		<-provider.connTimer.C
-		provider.connTimer.Reset(time.Duration(provider.RefreshConn))
-		tr.CancelRequest(req)
```

Comment:
Note that [CancelRequest](https://golang.org/pkg/net/http/#Transport.CancelRequest) has been deprecated and the new context functionality should be used instead.